File: grrp/src/grrp/accounts.py
"""Accounts: a password in front of an identity.

An account is *not* what makes a record credible, and it is worth being exact
about what it is and is not, because every system that has one eventually
starts treating it as the thing that matters.

An **identity** (:mod:`grrp.identity`) is a keypair.  It signs.  It is what
attribution and attestation rest on, it is portable, and it works with no
server in the picture at all.

An **account** is a name and a password that let a person reach an identity
through a browser on a server somebody else runs.  It is convenience and access
control.  It signs nothing.  Deleting every account here would not weaken a
single transition, because the signatures are over the keys and not over the
logins -- which is the property that makes this hostable without the host
becoming the authority (C10).

Registration is closed at the moment.  A signed-up stranger is not a party to
anything until somebody registers their acts, so open sign-up would add
accounts and no credibility; accounts are created by whoever runs the server,
with ``grrp account add``.
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import secrets
from dataclasses import dataclass
from pathlib import Path

from . import errors, identity, store

ACCOUNTS_DIR = ".grrp-accounts"

#: Whether the sign-in page will create an account for a stranger.  Closed for
#: now, and the page says so rather than showing a form that fails.
REGISTRATION_OPEN = False

# scrypt at the parameters the standard library documents as interactive.  The
# threat is somebody who obtained the account files, which is also somebody who
# obtained the private keys sitting next to them -- so this protects the
# password itself, on the assumption it is reused elsewhere, and nothing more.
_N, _R, _P, _LEN = 2**14, 8, 1, 32


@dataclass(frozen=True)
class Account:
    name: str
    party: str


def directory(root: Path) -> Path:
    return root / ACCOUNTS_DIR


def _path(root: Path, name: str) -> Path:
    return directory(root) / f"{name}.yaml"


def _b64(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).decode("ascii").rstrip("=")


def _unb64(text: str) -> bytes:
    return base64.urlsafe_b64decode(text + "=" * (-len(text) % 4))


def hash_password(password: str, salt: bytes | None = None) -> str:
    salt = salt or secrets.token_bytes(16)
    digest = hashlib.scrypt(
        password.encode("utf-8"), salt=salt, n=_N, r=_R, p=_P, dklen=_LEN
    )
    return f"scrypt${_N}${_R}${_P}${_b64(salt)}${_b64(digest)}"


def _matches(password: str, stored: str) -> bool:
    try:
        algorithm, n, r, p, salt, digest = stored.split("$")
        if algorithm != "scrypt":
            return False
        again = hashlib.scrypt(
            password.encode("utf-8"),
            salt=_unb64(salt),
            n=int(n),
            r=int(r),
            p=int(p),
            dklen=len(_unb64(digest)),
        )
    except Exception:  # noqa: BLE001 - an unreadable record is a failure to match
        return False
    return hmac.compare_digest(again, _unb64(digest))


def exists(root: Path, name: str) -> bool:
    return _path(root, name).is_file()


def listing(root: Path) -> list[Account]:
    """Every account on this server, by name.

    By name and never by anything else: any other ordering of participants
    ranks them, however it is labelled (C6).
    """
    if not directory(root).is_dir():
        return []
    out = []
    for path in sorted(directory(root).glob("*.yaml")):
        data = store.read_yaml(path)
        out.append(Account(data["name"], data["party"]))
    return out


def create(root: Path, name: str, password: str) -> Account:
    """Make an account, and the keypair it reaches, in one step."""
    name = name.strip().lower()
    if len(password) < 8:
        raise errors.Refused(
            "a password of at least 8 characters. This guards browser access on "
            "whatever machine runs the server; it does not guard the record, which "
            "is guarded by signatures."
        )
    if exists(root, name):
        raise errors.Refused(f"there is already an account named {name!r} here")

    try:
        who = identity.find(root, name)
    except errors.GrrpError:
        who = identity.create(root, name)

    directory(root).mkdir(parents=True, exist_ok=True)
    (directory(root) / ".gitignore").write_text(
        "# Password hashes are nobody's business but this server's.\n*\n",
        encoding="utf-8",
    )
    store.write_yaml(
        _path(root, name),
        {
            "name": name,
            "party": who.party,
            "password": hash_password(password),
            "created": store.now(),
            "note": "An account reaches an identity. The identity signs; this does not.",
        },
    )
    return Account(name, who.party)


def authenticate(root: Path, name: str, password: str) -> Account:
    """Check a name and password, or refuse without saying which was wrong."""
    name = name.strip().lower()
    path = _path(root, name)
    stored = store.read_yaml(path).get("password", "") if path.is_file() else ""
    # Hash regardless, so that a missing account and a wrong password take
    # about the same time and the page does not disclose who has an account.
    if not _matches(password, stored) or not stored:
        raise errors.Refused("that name and password do not go together")
    return Account(name, identity.find(root, name).party)


def set_password(root: Path, name: str, password: str) -> None:
    if not exists(root, name):
        raise errors.UnknownReference(f"no account named {name!r} here")
    data = store.read_yaml(_path(root, name))
    data["password"] = hash_password(password)
    store.write_yaml(_path(root, name), data)

File: web_application/packages/server/src/accounts.ts
/**
 * Accounts: a password in front of a public key.
 *
 * An account is **not** what makes a record credible, and being exact about
 * that is the whole discipline here, because every system with accounts
 * eventually starts treating them as the thing that matters.
 *
 *   An *identity* is a keypair. It signs. It is what attribution and
 *   attestation rest on, and it works with no server in the picture.
 *
 *   An *account* is a name and a password that let a person reach this server
 *   through a browser. It is access control. It signs nothing.
 *
 * The private key never arrives here. It is generated in the browser and stays
 * there; the server is told only the public half. That is not a policy anyone
 * has to trust -- there is no column it could be stored in.
 *
 * Registration is closed. A signed-up stranger is not a party to anything
 * until somebody registers their acts, so open sign-up would add accounts and
 * no credibility. Accounts are made by whoever runs the server.
 */

import { randomBytes, scrypt as scryptCallback, timingSafeEqual } from 'node:crypto'
import { promisify } from 'node:util'

import { eq } from 'drizzle-orm'

import type { Db } from './database.js'
import { users } from './db.js'

const scrypt = promisify(scryptCallback) as (
  password: string,
  salt: Buffer,
  keylen: number,
) => Promise<Buffer>

/** Whether the sign-in page will make an account for a stranger. */
export const REGISTRATION_OPEN = false

const KEYLEN = 32
export const NAME_PATTERN = /^[a-z0-9][a-z0-9._-]{0,31}$/

export class Refused extends Error {}

export interface Account {
  id: number
  name: string
  party: string
  displayName: string | null
}

export async function hashPassword(password: string): Promise<string> {
  const salt = randomBytes(16)
  const derived = await scrypt(password, salt, KEYLEN)
  return `scrypt$${salt.toString('base64url')}$${derived.toString('base64url')}`
}

async function matches(password: string, stored: string): Promise<boolean> {
  const [algorithm, salt, digest] = stored.split('$')
  if (algorithm !== 'scrypt' || !salt || !digest) return false
  const expected = Buffer.from(digest, 'base64url')
  const derived = await scrypt(password, Buffer.from(salt, 'base64url'), expected.length)
  return derived.length === expected.length && timingSafeEqual(derived, expected)
}

export async function createAccount(
  db: Db,
  input: { name: string; password: string; party: string; displayName?: string },
): Promise<Account> {
  const name = input.name.trim().toLowerCase()
  if (!NAME_PATTERN.test(name)) {
    throw new Refused(
      `${JSON.stringify(name)} will not do as a name: lower-case letters, digits, and . - _ , ` +
        'up to 32 characters. The name is for typing; the identity is the key, and nothing ' +
        'requires either to be your legal name.',
    )
  }
  if (input.password.length < 8) {
    throw new Refused(
      'a password of at least 8 characters. This guards browser access to this server; it ' +
        'does not guard the record, which is guarded by signatures.',
    )
  }
  if (!input.party.startsWith('key:ed25519:')) {
    throw new Refused('an account needs the public half of a keypair generated in your browser')
  }
  const row = {
    name,
    party: input.party,
    passwordHash: await hashPassword(input.password),
    displayName: input.displayName ?? null,
    createdAt: new Date().toISOString(),
  }
  try {
    const [created] = await db.insert(users).values(row).returning()
    return toAccount(created!)
  } catch (error) {
    if (String(error).includes('UNIQUE')) {
      throw new Refused(
        `there is already an account here under that name or that key. Two identities under ` +
          `one name would make the record say one party acted where two did.`,
      )
    }
    throw error
  }
}

/**
 * Check a name and password.
 *
 * Refuses with one sentence either way. Saying which half was wrong would
 * disclose who has an account here, which is a directory of participants by
 * another route.
 */
export async function authenticate(db: Db, name: string, password: string): Promise<Account> {
  const [row] = await db
    .select()
    .from(users)
    .where(eq(users.name, name.trim().toLowerCase()))
    .limit(1)
  // Hash regardless of whether the account exists, so a missing name and a
  // wrong password take about the same time.
  const stored = row?.passwordHash ?? `scrypt$${randomBytes(16).toString('base64url')}$${'A'.repeat(43)}`
  const ok = await matches(password, stored)
  if (!row || !ok) throw new Refused('that name and password do not go together')
  return toAccount(row)
}

export async function byName(db: Db, name: string): Promise<Account | null> {
  const [row] = await db.select().from(users).where(eq(users.name, name)).limit(1)
  return row ? toAccount(row) : null
}

function toAccount(row: typeof users.$inferSelect): Account {
  return { id: row.id, name: row.name, party: row.party, displayName: row.displayName }
}
